### PR TITLE
Avoid boxing floats when calling classify_float

### PR DESCRIPTION
**Pull request by alainfrisch**
The PR uses the new mechanism to specify float arguments to be passed in unboxed form when calling a C function in order to optimize calls to `Pervasives.classify_float`.

A new test ensures that boxing is avoiding.

On my machine, using the mingw 32-bit port (which forces to use the emulation of classify_float, cf floats.c), a loop such as:

``` ocaml
let () =
  let x = ref 1. in
  for i = 1 to 1000000000 do
    assert (classify_float !x = FP_normal);
    x := !x +. 1.
  done
```

is optimized as follow:

Before:  5.6s
After: 3.6s

And on linux 64 (in a vm):

Before: 8.7s
After: 6.6s

### CODE REVIEW

**Comment by mshinwell** [comment]
This looks good, but I think there should be a better way of making sure that the thing allocates zero.  (In Core_kernel, we have a separate C primitive for retrieving minor_words, although we always assume the result fits in the native OCaml integer width.)

**Comment by alainfrisch** [comment]
@mshinwell You mean, a better way to write the test, or some compiler support to guarantee that no allocation occur in a given piece of code (an attribute that triggers a warning)?

**Comment by jhjourdan** [comment]
This is maybe a naive remark, but wouldn't it be faster and safer to write this primitive in OCaml using the bitwise representation of floats and Int64.bits_of_float ?

**Comment by alainfrisch** [comment]
Another remark: using the fpclassify emulation in floats.c, even on Linux, brings even better performance:

Trunk: 8.7s
Branch: 6.6s
Branch, forcing to use the emulation: 4.0s

Does anyone know why fpclassify is slower than our emulation?

**Comment by alainfrisch** [comment]
@jhjourdan I doubt this could be made as fast, or even just avoid the allocation (even on 32-bit).

**Comment by mshinwell** [comment]
@alainfrisch For the moment, just a better test, I would have thought.  Although some kind of "no allocation allowed" attribute would be interesting.

**Comment by alainfrisch** [comment]
> just a better test

@mshinwell   Note that this PR only reuses the technique currently used in the testsuite, in several places, to detect the absence of allocation.

**Comment by alainfrisch** [comment]
@bobot I've no big plan for other functions, I had just been bitten in practice by the bad performance of classify_float and wanted to address it.  I don't see any drawback in providing unboxed variants of other functions when this is possible.

**Comment by alainfrisch** [comment]
While not strictly related to this discussion, it would be interesting if others could confirm that using our emulation instead of fpclassify (cf floats.c) beats the libc version (and it would be even better if someone had an explanation!).

**Comment by ghost** [comment]
> In Core_kernel, we have a separate C primitive for retrieving minor_words, although we always assume the result fits in the native OCaml integer width.

@mshinwell, that's no problem, now we can do:

``` ocaml
external minor_words : unit -> (int64 [@unboxed]) = "blah" "blah_unboxed"
```

**Comment by jhjourdan** [comment]
@alainfrisch, The real probelm is that Int64.bits_of_float uses a foreign C call, which seems awkward, given that it is really just identity and that most operations on Int64 are compiled directly to assembly (maybe optimizing should be a good idea).

Using a hack in order to avoid this foreign call (works only on 64bits):

```
type hack = { pad: float; mutable x :float }
let data = { pad = 0.; x = 0. }

let fpclass f =
  data.x <- f;
  let i = Obj.magic data in (* Simulate boxing of an Int64 *)
  let i = Int64.add i i in  (* Mask sign bit *)
  match Int64.shift_right i 53 with
  | 0L -> if i = 0L then FP_zero else FP_subnormal
  | -1L -> if i = 0xFFE0000000000000L then FP_infinite else FP_nan
  | _ -> FP_normal

let () =
  let x = ref 1. in
  for i = 1 to 1000000000 do
    assert (fpclass !x = FP_normal);
    x := !x +. 1.
  done
```

This gives really good performances (3.49s. vs. 7.98s. in the current trunk).

**Comment by jhjourdan** [comment]
And when forcing inlining of fpclass, this goes down to 1.6s.

File: testsuite/tests/float-unboxing/float_subst_boxed_number.ml [hunk 1]
```diff
@@ -38,6 +38,20 @@ module PR_6770 = struct
...
 
 
+let check_noalloc name f =
+  let a0 = Gc.allocated_bytes () in
+  let a1 = Gc.allocated_bytes () in
+  let _x = f () in
+  let a2 = Gc.allocated_bytes () in
+  let alloc = (a2 -. 2. *. a1 +. a0) in
+
+  (* is there a better to test whether we run in native code? *)
```
Comment (chambart) [inline testsuite/tests/float-unboxing/float_subst_boxed_number.ml:48]: Yes, Dynlink.is_native

**Comment by alainfrisch** [comment]
@jhjourdan This is very interesting!  Do you think good performances could be achieved on 32-bit as well?  I assume that the OCaml version is significantly slower in bytecode, which might be a bit problematic.  

Anyway, it seems the C version would still be a bit faster, even compared to the inlined OCaml version on 64-bit.

**Comment by jhjourdan** [comment]
@alainfrisch 64bits computations in 32bits end up in library calls, so that it will not be competitive, and yes, the performance in bytecode is bad.

Apart from that, I still think a optimized version of bits_of_float is worth it, given that it should be trivial to implement.

**Comment by xavierleroy** [comment]
@jhjourdan : ocamlopt native support for bits_of_float/float_of_bits involves a move between a FP register and an integer register.  This is supported by some processors (e.g. x86-SSE2, ARM) but for others (x86-x87, PowerPC) a transfer through memory is required.  Plus, this means pushing two new C-- operators down every ocamlopt back-end.  Hardly worth the effort, I think.

**Comment by alainfrisch** [comment]
@xavierleroy Do you have an idea why the emulation is faster than the libc version?  Could it be explained only by the avoided function call?

**Comment by xavierleroy** [comment]
@alainfrisch : the cost of calling fpclassify (it's not an inline function in Glibc), yes, plus the cost of converting fpclassify's return codes to the corresponding OCaml constant constructors.  

Note that if we make the emulation the default, a 64-bit version can be written that would be even faster on 64-bit platforms.

**Comment by alainfrisch** [comment]
@xavierleroy   That would be great and even increase the gain compared to the 64-bit version in OCaml.  So it seems we are going to stick to an external C primitive for Pervasives.classify_float.  Is anyone against merging the current PR?

File: byterun/floats.c [hunk 2]
```diff
@@ -451,12 +451,12 @@ CAMLprim value caml_float_compare(value vf, value vg)
 
 enum { FP_normal, FP_subnormal, FP_zero, FP_infinite, FP_nan };
 
-CAMLprim value caml_classify_float(value vd)
+CAMLprim value caml_classify_float_unboxed(double vd)
```
Comment (ghost) [inline byterun/floats.c]: I suppose the `CAMLprim` is not necessary here? I didn't put it this #277

**Comment by xavierleroy** [comment]
Commit [trunk e7f339e] improves the performance of classify_float, as discussed here.

**Pull request merged**